### PR TITLE
[Divulgence pruning] Added `prune_all_divulged_contracts` to PruneRequest [DPP-534]

### PR DESCRIPTION
**Pull request by tudor-da**
This PR enriches the `PruningRequest` with an additional flag `prune_all_divulged_contracts`. 
* The flag is passed to the `WriteParticipantPruningService` (v1 and v2)

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by meiersi-da — COMMENTED** [review]
Looks good up to two concerns:
1. The `sandbox-classic` codebase must return `NOT_IMPLEMENTED` when the `prune_all_divulged_contracts` flag is set. I'd expect this PR to make that change, or is this the wrong place?
2. Do the KV ledgers support `prune_all_divulged_contracts` as of this PR? If not, I'd suggest for them to return  `NOT_IMPLEMENTED` when the `prune_all_divulged_contracts` flag is set.

This way this PR leaves the codebase in a consistent state, and later changes can replace the `NOT_IMPLEMENTED` errors with their actual implementation.

What do you think?

**Comment by tudor-da** [comment]
> Looks good up to two concerns:
> 
>     1. The `sandbox-classic` codebase must return `NOT_IMPLEMENTED` when the `prune_all_divulged_contracts` flag is set. I'd expect this PR to make that change, or is this the wrong place?
> 
>     2. Do the KV ledgers support `prune_all_divulged_contracts` as of this PR? If not, I'd suggest for them to return  `NOT_IMPLEMENTED` when the `prune_all_divulged_contracts` flag is set.
> 
> 
> This way this PR leaves the codebase in a consistent state, and later changes can replace the `NOT_IMPLEMENTED` errors with their actual implementation.
> 
> What do you think?

Afaict, sandbox-classic already returns `UNIMPLEMENTED` (see [here](https://github.com/digital-asset/daml/blob/6bd297a628ee364b668a01d3a2150d7e8709b008/ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/LedgerBackedWriteService.scala#L101)). For KV ledgers, I didn't see an issue if the interface slightly inconsistent for between the commits of this epic, but the change is small, I'll amend.

**Review by meiersi-da — APPROVED** [review]
Thx @tudor-da . Sorry for overlooking the `Status.UNIMPLEMENTED` for `sandbox-classic`.

LGTM.

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteParticipantPruningService.scala [hunk 1]
```diff
@@ -35,10 +35,18 @@ trait WriteParticipantPruningService {
...
+    * @param submissionId The submission id.
+    * @param pruneAllDivulgedContracts If set, instruct the ledger to prune all immediate and retroactive divulged contracts
+    *                                  created before `pruneUpToInclusive` independent of whether they were archived before
+    *                                  `pruneUpToInclusive`.
+    * @return The pruning result.
     */
   def prune(
       pruneUpToInclusive: Offset,
       submissionId: Ref.SubmissionId,
+      pruneAllDivulgedContracts: Boolean,
```
Comment (tudor-da) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteParticipantPruningService.scala:49]: CC @oliverse-da Coming your way

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala [hunk 2]
```diff
@@ -35,10 +35,17 @@ trait WriteParticipantPruningService {
...
+    * @param submissionId The submission id.
+    * @param pruneAllDivulgedContracts If set, instruct the ledger to prune all immediate and retroactive divulged contracts
+    *                                  created before `pruneUpToInclusive` independent of whether they were archived before
+    *                                  `pruneUpToInclusive`.
+    * @return The pruning result.
     */
   def prune(
       pruneUpToInclusive: Offset,
       submissionId: Ref.SubmissionId,
+      pruneAllDivulgedContracts: Boolean,
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:49]: Can you do this without modifying the v1 WriteService? We'd prefer to keep this stable, especially as it's going away soon.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala [hunk 3]
```diff
@@ -46,7 +47,8 @@ class ParticipantPruningIT extends LedgerTestSuite {
   )(implicit ec => { case Participants(Participant(participant)) =>
     for {
       cannotPruneNonHexOffset <- participant
-        .prune("cofefe", attempts = 1)
+        // TODO Divulgence pruning: Change to `true` once all divulgence pruning is implemented
+        .prune("cofefe", attempts = 1, pruneAllDivulgedContracts = false)
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala:51]:
```suggestion
        .prune("covfefe", attempts = 1, pruneAllDivulgedContracts = false)
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:49]: The `KeyValueParticipantStateWriter` still implements this interface, hence we need it for implementing full divulgence pruning on KV ledgers. I'll have a look, but not sure if I can circumvent this.

Comment (SamirTalwar-DA) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:49]: How do you feel about waiting a couple of days until work is done to migrate `KeyValueParticipantStateWriter` to `v2.WriteService`?

Comment (tudor-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:49]: Unfortunately, this implementation needs to catch this week's snapshot release

Comment (SamirTalwar-DA) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:49]: No worries, it happens.

**Review by SamirTalwar-DA — APPROVED** [review]

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala (hunk 3 above)
Comment (tudor-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ParticipantPruningIT.scala]: Done

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto [hunk 4]
```diff
@@ -39,12 +39,29 @@ service ParticipantPruningService {
...
+  //   1. All normal and divulged contracts that have been archived before
+  //   `prune_up_to`.
+  //   2. All transaction events and completions before `prune_up_to`
   string prune_up_to = 1;
 
   // Unique submission identifier.
   // Optional, defaults to a random identifier, used for logging.
   string submission_id = 2;
+
+  // Prune all immediate and retroactive divulged contracts created before `prune_up_to`
```
Comment (mziolekda) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto:53]:
```suggestion
  // Prune all immediately and retroactively divulged contracts created before `prune_up_to`
```

**Review by hubert-da — APPROVED** [review]

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala [hunk 5]
```diff
@@ -35,10 +35,17 @@ trait WriteParticipantPruningService {
     * step upon successful completion of each prune call. To reach eventual consistency upon failures, be sure
     * to return ParticipantPruned even if the specified offset has already been pruned to allow ledger api server
     * index pruning to proceed in case of an earlier failure.
+    *
+    * @param pruneUpToInclusive The offset up to which contracts should be pruned.
+    * @param submissionId The submission id.
+    * @param pruneAllDivulgedContracts If set, instruct the ledger to prune all immediate and retroactive divulged contracts
```
Comment (mziolekda) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala:41]:
```suggestion
    * @param pruneAllDivulgedContracts If set, instruct the ledger to prune all immediately and retroactively divulged contracts
```

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteParticipantPruningService.scala [hunk 6]
```diff
@@ -35,10 +35,18 @@ trait WriteParticipantPruningService {
     * step upon successful completion of each prune call. To reach eventual consistency upon failures, be sure
     * to return ParticipantPruned even if the specified offset has already been pruned to allow ledger api server
     * index pruning to proceed in case of an earlier failure.
+    *
+    * @param pruneUpToInclusive The offset up to which contracts should be pruned.
+    * @param submissionId The submission id.
+    * @param pruneAllDivulgedContracts If set, instruct the ledger to prune all immediate and retroactive divulged contracts
```
Comment (mziolekda) [inline ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteParticipantPruningService.scala:41]:
```suggestion
    * @param pruneAllDivulgedContracts If set, instruct the ledger to prune all immediately and retroactively divulged contracts
```

**Review by mziolekda — APPROVED** [review]

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala [hunk 7]
```diff
@@ -706,17 +706,28 @@ private[testtool] final class ParticipantTestContext private[participant] (
...
+        )
       },
     )
 
-  def prune(pruneUpTo: LedgerOffset, attempts: Int = 10): Future[PruneResponse] =
-    prune(pruneUpTo.getAbsolute, attempts)
+  def prune(
+      pruneUpTo: LedgerOffset,
+      attempts: Int = 10,
+      // TODO Divulgence pruning: Change default to `true` once all divulgence pruning is implemented
```
Comment (fabiotudone-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala:727]: If this is tracked, I think you can remove this and other occurrences.

**Review by fabiotudone-da — APPROVED** [review]

File: ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteParticipantPruningService.scala (hunk 1 above)
Comment (oliverse-da) [reply ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteParticipantPruningService.scala:49]: fyi @matthiasS-da - merged now so will be in this week's daml snapshot upgrade.

**Pull request merged**